daemon/cmd/policy.go

```diff
@@ -58,20 +63,92 @@ func (d *Daemon) initPolicy(
 	}
 	d.datapathRegenTrigger = rt
 
-	d.policy = policy.NewPolicyRepository(d.identityAllocator,
-		d.identityAllocator.GetIdentityCache(),
-		certManager,
-		secretManager,
+	d.monitorAgent.RegisterNewConsumer(authManager)
+	return nil
+}
+
+type policyParams struct {
+	cell.In
+
+	Lifecycle       hive.Lifecycle
+	EndpointManager endpointmanager.EndpointManager
+	CertManager     certificatemanager.CertificateManager
+	SecretManager   certificatemanager.SecretManager
+	Datapath        datapath.Datapath
+	CacheStatus     k8s.CacheStatus
+}
+
+type policyOut struct {
+	cell.Out
+
+	IdentityAllocator CachingIdentityAllocator
+	Repository        *policy.Repository
+	Updater           *policy.Updater
+	IPCache           *ipcache.IPCache
+}
+
+// newPolicyTrifecta instantiates CachingIdentityAllocator, Repository and IPCache.
+//
+// The three have a circular dependency on each other and therefore require
+// special care.
+func newPolicyTrifecta(params policyParams) (policyOut, error) {
+	iao := &identityAllocatorOwner{}
+	idAlloc := &cachingIdentityAllocator{
+		cache.NewCachingIdentityAllocator(iao),
+		nil,
+	}
+
+	iao.policy = policy.NewStoppedPolicyRepository(
+		idAlloc,
+		idAlloc.GetIdentityCache(),
+		params.CertManager,
+		params.SecretManager,
 	)
-	d.policy.SetEnvoyRulesFunc(envoy.GetEnvoyHTTPRules)
-	d.policyUpdater, err = policy.NewUpdater(d.policy, epMgr)
+	iao.policy.SetEnvoyRulesFunc(envoy.GetEnvoyHTTPRules)
+
+	policyUpdater, err := policy.NewUpdater(iao.policy, params.EndpointManager)
 	if err != nil {
-		return fmt.Errorf("failed to create policy update trigger: %w", err)
+		return policyOut{}, fmt.Errorf("failed to create policy update trigger: %w", err)
 	}
+	iao.policyUpdater = policyUpdater
 
-	d.monitorAgent.RegisterNewConsumer(authManager)
+	ctx, cancel := context.WithCancel(context.Background())
 
-	return nil
+	ipc := ipcache.NewIPCache(&ipcache.Configuration{
+		Context:           ctx,
+		IdentityAllocator: idAlloc,
+		PolicyHandler:     iao.policy.GetSelectorCache(),
+		DatapathHandler:   params.EndpointManager,
+		NodeHandler:       params.Datapath.Node(),
+		CacheStatus:       params.CacheStatus,
+	})
+	idAlloc.ipcache = ipc
+
+	params.Lifecycle.Append(hive.Hook{
+		OnStart: func(hc hive.HookContext) error {
+			iao.policy.Start()
+			return nil
+		},
+		OnStop: func(hc hive.HookContext) error {
+			// Preserve the order of shutdown but still propagate the error
+			// to hive.
+			err := ipc.Shutdown()
+			policyUpdater.Shutdown()
+			idAlloc.Close()
+
+			// Cancel after shutting down IPCache.
+			cancel()
```

Comment:
This was previously tied to the Daemon context and closed *before* `ipc.Shutdown()`, related to https://github.com/cilium/cilium/pull/21676. At the very least I think we should move this context up to close first -- basically this gives the early signal to start shutting down some controllers in `pkg/ipcache` which `Shutdown()` will want to wait on anyway.

Reply:
(arguably we could maybe get rid of the `Context` here now if we're not trying to converge all the contexts onto the root tree of `Context`s any more, especially given that there's already an explicit `Shutdown()` call going on here, but that's more of an IPCache cleanup rather than what you're trying to achieve in this PR, so I'm happy to defer that work until later. For now I just would like to prevent unit test failures due to race conditions between goroutines shutting down)

Reply:
> I think we should move this context up to close first -- basically this gives the early signal to start shutting down some controllers in `pkg/ipcache` which `Shutdown()` will want to wait on anyway.

Ack, that's what I had in the beginning. Thanks for the context, I'll change it back.

Reply:
(cc @gandro , I missed that you had conflicting suggestions but I do think that the cancel() needs to happen first here)